ts/util/reactions.ts

```diff
@@ -0,0 +1,249 @@
+
+  window.log.info(
+    `${sender} ${reaction.action === 0 ? 'added' : 'removed'} a ${reaction.emoji} reaction`
+  );
+
+  switch (reaction.action) {
+    case SignalService.DataMessage.Reaction.Action.REACT:
+      if (senders.includes(sender) && details.senders[sender] !== '') {
+        window?.log?.info(
+          'Received duplicate message reaction. Dropping it. id:',
+          details.senders[sender]
+        );
+        return;
+      }
+      details.senders[sender] = messageId ?? '';
+      break;
+    case SignalService.DataMessage.Reaction.Action.REMOVE:
+    default:
+      if (senders.length > 0) {
+        if (senders.indexOf(sender) >= 0) {
+          // tslint:disable-next-line: no-dynamic-delete
+          delete details.senders[sender];
+        }
+      }
+  }
+
+  const count = Object.keys(details.senders).length;
+  if (count > 0) {
+    reacts[reaction.emoji].count = count;
+    reacts[reaction.emoji].senders = details.senders;
+
+    // sorting for open groups convos is handled by SOGS
+    if (!isOpenGroup && details && details.index === undefined) {
+      reacts[reaction.emoji].index = originalMessage.get('reactsIndex') ?? 0;
+      originalMessage.set('reactsIndex', (originalMessage.get('reactsIndex') ?? 0) + 1);
+    }
+  } else {
+    // tslint:disable-next-line: no-dynamic-delete
+    delete reacts[reaction.emoji];
+  }
+
+  originalMessage.set({
+    reacts: !isEmpty(reacts) ? reacts : undefined,
+  });
+
+  await originalMessage.commit();
+  return originalMessage;
+};
+
+/**
+ * Handle all updates to messages reactions from the SOGS API
+ */
+export const handleOpenGroupMessageReactions = async (
+  reactions: OpenGroupReactionList,
+  serverId: number
+) => {
+  const originalMessage = await Data.getMessageByServerId(serverId);
+  if (!originalMessage) {
+    window?.log?.warn(`Cannot find the original reacted message ${serverId}.`);
+    return;
+  }
+
+  if (isEmpty(reactions)) {
+    if (originalMessage.get('reacts')) {
+      originalMessage.set({
+        reacts: undefined,
+      });
+    }
+  } else {
+    const reacts: ReactionList = {};
+    Object.keys(reactions).forEach(key => {
+      const emoji = decodeURI(key);
+      const senders: Record<string, string> = {};
+      reactions[key].reactors.forEach(reactor => {
+        senders[reactor] = String(serverId);
+      });
+      reacts[emoji] = { count: reactions[key].count, index: reactions[key].index, senders };
+    });
+
+    originalMessage.set({
+      reacts,
+    });
+  }
+
+  await originalMessage.commit();
+  return originalMessage;
+};
+
+export const updateRecentReactions = async (reactions: Array<string>, newReaction: string) => {
+  window?.log?.info('updating recent reactions with', newReaction);
+  const recentReactions = new RecentReactions(reactions);
+  const foundIndex = recentReactions.items.indexOf(newReaction);
+  if (foundIndex >= 0) {
+    if (foundIndex === 0) {
+      return;
+    }
```

Comment:
it's just a preference but I like the early return ifs
breaks down a bit 
```
if (foundIndex === 0) {
      return;
    }
 if (foundIndex > 0) {
...
}
```

Reply:
done